// brunch/lib/fs_utils/asset.js
// Generated by CoffeeScript 1.9.1
(function() {
  'use strict';
  var Asset, common, debug, getAssetDirectory, sysPath;

  debug = require('debug')('brunch:asset');

  sysPath = require('path');

  common = require('./common');

  getAssetDirectory = function(path, convention) {
    var split;
    split = path.split(common.sep);
    return split.map(function(part, index) {
      return split.slice(0, index).concat([part, '']).join(common.sep);
    }).filter(convention)[0];
  };

  module.exports = Asset = (function() {
    function Asset(path1, config) {
      var directory;
      this.path = path1;
      directory = getAssetDirectory(this.path, config._normalized.conventions.assets);
      this.relativePath = sysPath.relative(directory, this.path);
      this.destinationPath = sysPath.join(config.paths["public"], this.relativePath);
      debug("Initializing fs_utils.Asset %s", JSON.stringify({
        path: this.path,
        directory: directory,
        relativePath: this.relativePath,
        destinationPath: this.destinationPath
      }));
      this.error = null;
      this.copyTime = null;
      Object.seal(this);
    }

    Asset.prototype.copy = function(callback) {
      return common.copy(this.path, this.destinationPath, (function(_this) {
        return function(error) {
          var err;
          if (error != null) {
            err = new Error(error);
            err.code = 'Copying';
            _this.error = err;
          } else {
            _this.error = null;
          }
          _this.copyTime = Date.now();
          return callback(_this.error);
        };
      })(this));
    };

    return Asset;

  })();

}).call(this);
